codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java

```diff
@@ -705,34 +699,25 @@ private void headerSizeExceeded() throws Http2Exception {
          *            block. In that case, the buffer is used directly without copying.
          */
         final void addFragment(ByteBuf fragment, ByteBufAllocator alloc, boolean endOfHeaders) throws Http2Exception {
-            if (headerBlock == null) {
-                if (fragment.readableBytes() > headersDecoder.configuration().maxHeaderListSizeGoAway()) {
-                    headerSizeExceeded();
-                }
-                if (endOfHeaders) {
-                    // Optimization - don't bother copying, just use the buffer as-is. Need
-                    // to retain since we release when the header block is built.
-                    headerBlock = fragment.retain();
+            if (unprocessedHeaderBlock == null) {
+                unprocessedHeaderBlock = alloc.compositeBuffer();
+            }
+            // Need to retain since CompositeByteBuf expects to own the reference.
+            unprocessedHeaderBlock.addComponent(fragment.retain());
+            Http2Headers headers;
+            boolean didThrow = true;
+            try {
+                if (!continuation) {
+                    headers = headersDecoder.decodeHeadersStart(streamId, unprocessedHeaderBlock, endOfHeaders);
+                    continuation = true;
                 } else {
-                    headerBlock = alloc.buffer(fragment.readableBytes());
-                    headerBlock.writeBytes(fragment);
+                    headers = headersDecoder.decodeHeadersContinue(unprocessedHeaderBlock, endOfHeaders);
+                }
+                didThrow = false;
+            } finally {
```

Comment:
On re-throwing you need to down-cast to RuntimeException, Error, and Http2Exception, or you need to use a Java language hack like PlatformDependent.throwException. I know Netty likes to use the PlatformDependent style, but it has bitten me because it is completely unchecked (it is possible to accidentally throw Exceptions). Java 7 multicatch would be a solution as long as you assume no throwables other than Exception and Error are thrown.

Reply:
TIL. So much easier in Scala where exceptions are unchecked, until there are exceptions being thrown and you have no idea which ones to expect. 😉